File: lib/languages/zig/guards.zig
// Zig doesn't have dedicated guard clauses like Gleam or Rust.
// Instead, it uses early returns, switch statements, and if conditions.

// Early returns as guards
fn processPositive(x: i32) !i32 {
    if (x <= 0) {
        return error.NegativeValue;
    }
    
    // Only reached if x > 0
    return x * 2;
}

// Multiple guard conditions
fn validateUser(user: User) !void {
    if (user.name.len == 0) {
        return error.EmptyName;
    }
    
    if (user.age < 18) {
        return error.TooYoung;
    }
    
    if (user.email.len == 0) {
        return error.NoEmail;
    }
    
    // Only reached if all validations pass
    std.debug.print("User is valid\n", .{});
}

// Switch with range conditions (similar to guards)
fn describe(x: i32) []const u8 {
    return switch (true) {
        x == 0 => "Zero",
        x > 0 and x < 10 => "Small positive",
        x >= 10 and x < 100 => "Medium positive",
        x >= 100 => "Large positive",
        x < 0 and x > -10 => "Small negative",
        x <= -10 and x > -100 => "Medium negative",
        else => "Large negative",
    };
}

// If-else chains with conditions
fn processPair(a: i32, b: i32) []const u8 {
    if (a == b) {
        return "Equal";
    } else if (a > b) {
        return "First is larger";
    } else {
        return "Second is larger";
    }
}

// Unwrapping optionals with guards
fn processOptional(value: ?i32) []const u8 {
    if (value) |v| {
        if (v > 0) {
            return "Positive";
        } else if (v < 0) {
            return "Negative";
        } else {
            return "Zero";
        }
    } else {
        return "No value";
    }
}

// Helper type for examples
const User = struct {
    name: []const u8,
    age: u32,
    email: []const u8,
};

File: lib/languages/python/guards.py
# Early returns as guards
def process_positive(x):
   if x <= 0:
       return "Must be positive"
   
   # Only reached if x > 0
   return x * 2

# Multiple guard conditions
def validate_user(user):
   if not user:
       return "User is required"
   
   if not user.get("name"):
       return "Name is required"
   
   if not user.get("email"):
       return "Email is required"
   
   if user.get("age", 0) < 18:
       return "Must be 18 or older"
   
   # Only reached if all validations pass
   return "User is valid"

# Guards with exceptions
def divide(a, b):
   if b == 0:
       raise ValueError("Cannot divide by zero")
   
   return a / b

# Type checking guards
def process_string(value):
   if not isinstance(value, str):
       raise TypeError("Expected a string")
   
   return value.upper()

# Guards with assertions
def calculate_square_root(x):
   assert x >= 0, "Input must be non-negative"
   
   return x ** 0.5

# Decorators as guards
def require_positive(func):
   def wrapper(x, *args, **kwargs):
       if x <= 0:
           raise ValueError("Argument must be positive")
       return func(x, *args, **kwargs)
   return wrapper

@require_positive
def calculate_log(x):
   import math
   return math.log(x)

